docs/synapse/userguides/storm_adv_control.rstorm

```diff
@@ -24,14 +24,151 @@ See also the follwing User Guide and reference sections for additional informati
+
+.. _storm-adv-debug:
+
+Storm Debugging Tips
+++++++++++++++++++++
+
+A few helpful tips when writing and debugging advanced Storm:
+
+**Be aware of your pipeline.**
+
+That is, understand what is in your current working set at any point in your query. A significant part of Storm troubleshooting comes down to figuring out that the current working set is not what you think it is.
+
+**Operations may execute multiple times.**
+
+Because each node passes through each operation in a Storm query individually, operations execute more than once (typically once for each node in the pipeline as it passes through that operation). This includes control flow operations, such as for loops! If you don't account for this behavior with control flow operations in particular, it can result in behavior such as:
+
+ - An exponentially increasing working set (if each node passing through an operation generates multiple results, and the results are not deduplicated / uniq'ed appropriately).
+ - A variable that is set by an operation being consistently changed (re-set) for each node passing through the operation (commonly resulting in "last node wins").
+ - A variable that **fails** to be set for a node that does **not** pass through the operation where the variable is assigned (resulting in a ``NoSuchVar`` error).
+ 
+**Use subqueries...but understand how they work.**
+ 
+Unlike most Storm operations and commands, subqueries **do not consume** nodes - by default, what goes into a subquery comes out of a subquery, regardless of what happens inside the subquery itself. This means you can use subqueries (see :ref:`storm-ref-subquery`) with advanced Storm to isolate certain operations and keep the "primary" nodes passing through the Storm pipeline consistent.
+
+That said, a node still has to pass **into** a subquery for the Storm inside a subquery to run. If your subquery **fails** to execute, it may be because nothing is going in to it.
+
+**Start small and add to your Storm incrementally.**
+
+It's generally easier to verify that smaller Storm queries execute correctly and then build on that code than to try and write a complex query all at once and try to figure out where things aren't working.
+
+**As with all debugging, print statements are your friend.**
+
+Scatter ``$lib.print()`` statements (see :ref:`stormlibs-lib-print`) generously throughout your Storm during testing. You can of course print message strings:
+
+::
+  
+  $lib.print("Hey! This worked!")
+
+You can print the value of a variable, to check its value at a given point in your query:
+
+::
+  
+  inet:ipv4=1.2.3.4
+  $asn=:asn
+  $lib.print($asn)
+
+You can also print values associated with the node(s) in the current working set, using the various methods associated with the ``$node`` Storm type. (See :ref:`storm-adv-methods` for a user-focused introduction to methods, or :ref:`stormprims-storm-node-f527` in the detailed Storm Libraries / Storm Types documentation for a more technical discussion.)
```

Comment:
just checking, is `stormprims-storm-node-f527` a valid reference? the "f527" looks odd

Reply:
It is a valid ref / link, though I agree it's weird. Maybe @vEpiphyte can clarify why it's wonky.

Reply:
It's a magic number used as a delimiter since we can have some label in the storm prims docs otherwise.

Reply:
:1st_place_medal: